R/submit_ipr.R

```diff
@@ -1,145 +1,417 @@
-    message <- paste0("POST error with sequence: ", fasta_header, "\n",
-                      "Status code: ", as.integer(res_post$status), "\n")
+
+  status <- ifelse(as.integer(res_post$status) == 200L, "running", "failed")
+  if (status == "failed") {
+    message <- paste0(
+      "POST error with sequence: ", fasta_header, "\n",
+      "Status code: ", as.integer(res_post$status), "\n"
+    )
     warning(message)
-    return(NULL)
-  }
-  else {
-    cat("POST status: ", as.integer(res_post$status), "\n", sep = "")
+    .add_job_status_table(basename(path2seq), job_id, status)
+    return("failed")
+  } else {
+    .add_job_status_table(basename(path2seq), job_id, status)
+    return(job_id)
   }
-  job_id <- rawToChar(res_post$content)
-  cat("job_id: ", job_id, "\n", sep = "")
+}
+
 
-  ### GET
+.poll_job <- function(job_code) {
+  cat("Polling job: ", job_code, "\n", sep = "")
   url_get_base <- "https://www.ebi.ac.uk/Tools/services/rest/iprscan5/result/"
-  url_get <- paste0(url_get_base, job_id, "/tsv")
+  url_get <- paste0(url_get_base, job_code, "/tsv")
   header_get <- "Accept: text/tab-separated-values"
+  res_get <- httr::GET(url_get, httr::add_headers(header_get))
 
-  status <- -1L
-  n_polls <- 1L
-  # Try poll 30 times OR until status is complete
-  while (status != 200L && n_polls < 180L) {
-    res_get <- httr::GET(url_get, httr::add_headers(header_get))
-    status <- as.integer(res_get$status_code)
-    cat("GET poll #", n_polls, "\t", "Status: ", status, "\n", sep = "")
-    n_polls <- n_polls + 1L
-    Sys.sleep(5L)
-  }
-  if (status != 200L) {
-    message <- paste0("Failed to retrieve job results for <",
-                      job_id,
-                      ">",
-                      "\n",
-                      "Job exceeded 30 minutes runtime.")
-    warning(message)
+  status <- as.integer(res_get$status_code)
+  status <- ifelse(status == 200L, "completed", "running")
+
+  # update N polls & latest poll time
+  tb_status <<- tb_status %>%
+    dplyr::mutate(polls = ifelse(job_id == job_code, polls + 1, polls)) %>%
+    dplyr::mutate(t_latest_poll = ifelse(job_id == job_code, as.POSIXct(Sys.time()), t_latest_poll))
+
+  return(status)
+}
+
+.update_status_table <- function(job_code, status_update) {
+  # change status for a given job_id row
+  tb_status <<- tb_status %>%
+    dplyr::mutate(status = ifelse(job_id == job_code, status_update, status))
+  cat("tb_status from .update_status_table()\n")
+  print(tb_status)
+}
+
+.lookup_input_file <- function(job_code) {
+  filename <- tb_status %>%
+    dplyr::filter(job_id == job_code) %>%
+    dplyr::pull(input_file)
+  return(filename)
+}
+
+.lookup_polls <- function(job_code) {
+  cat("Looking up poll count for: ", job_code, "\n", sep = "")
+  poll_count <- tb_status %>%
+    dplyr::filter(job_id == job_code) %>%
+    dplyr::pull(polls)
+  return(poll_count)
+}
+
+.get_running_jobs <- function() {
+  print(tb_status)
+  n_running <- tryCatch(
```

Comment:
Never used `tryCatch` before -- nice! Is it working well?

Reply:
Yeah! I'm new to using it, too. 

In this case, .get_running_jobs() attempts to read the status tibble and count how many jobs are running, but if there's an error in the process it returns `0`. `0` is important for the first iteration of job submission/polling loop in the line `n_available <- 30L - .get_running_jobs() # .get_running_jobs() return N running jobs`. Where I need this value to return 0 before any jobs have been added to the status tibble (because this is the first iteration). 

The purpose is to handle instances where `.get_running_jobs()` is called before the status tibble has any values. Therefore, the status column is empty and filtering a NULL vector raises an error.

Another instance is when I set the initial condition for iteratively processing multiple jobs.

`while ((n_seqs > 0L) || ((.get_running_jobs()) > 0L)) `

I'm realizing it would probably help to add a few comments in the code about this.